### ISSUE TITLE
Add thermal storage

### ISSUE THREAD

**Issue by stroitzsch** [TUMCREATE-ESTL/cobmo#1]
## Problem:

Thermal storage model is needed to study the increase in demand side flexibility for buildings with such storage systems.

## Notes:

- Thermal storage is inserted between chiller plant and AHU / TU on zone-level, i.e., there is a heat transfer between chiller plant and thermal storage as well as between thermal storage and zone.
- Thermal storage is defined on building-level, i.e., there is only one thermal storage per building.

**Comment by stroitzsch** [comment]
As discussed:

- [ ] Thermal storage doesn't need constant:
https://github.com/TUMCREATE-ESTL/cobmo/blob/fb84a27177cb74b6404fe3d1dc0393f62ce6505e/cobmo/building.py#L94

- [ ] Thermal storage shall be defined on building level instead of zone level:
https://github.com/TUMCREATE-ESTL/cobmo/blob/fb84a27177cb74b6404fe3d1dc0393f62ce6505e/cobmo/building.py#L118-L121
https://github.com/TUMCREATE-ESTL/cobmo/blob/fb84a27177cb74b6404fe3d1dc0393f62ce6505e/data/building_zone_types.csv#L1-L3

(Just testing the code quoting capabilities)

**Comment by tommasomiori** [comment]
I still see the problem in my last commit (_569996438bcf1e0aa040e425614b3ea070fcc0dc_) with the lack of a constant for storage. 
@stroitzsch can you give me a hint?

**Comment by tommasomiori** [comment]
--> when adding the storage to the scenarios, add it to the default.
--> boolean error:  solving it with a sentence "_something_ IF _condition_"

**Comment by stroitzsch** [comment]
As discussed:
```
(self.building_zones['zone_name'] + '_storage_to_zone_heat_power') if (
                    (self.building_scenarios['building_storage_type'][0] == 'sensible_thermal_storage_default')  #
                    | (self.building_scenarios['building_storage_type'][0] == 'latent_thermal_storage_default')  #
                ) else None)
```

**Comment by tommasomiori** [comment]
’ code line

**Comment by stroitzsch** [comment]
## Nomenclature Comments

- [ ] `_tank_temperature` :arrow_right: `_sensible_thermal_storage_state_of_charge` (actually will need several states, right?)
https://github.com/TUMCREATE-ESTL/cobmo/compare/feature/thermal_storage#diff-bbc55971cd554ab8d16f6643b1f6b422R127
- [ ] `_pcm_state_of_charge` :arrow_right: `_latent_thermal_storage_state_of_charge`
https://github.com/TUMCREATE-ESTL/cobmo/compare/feature/thermal_storage#diff-bbc55971cd554ab8d16f6643b1f6b422R130
- [ ] `_batteries_state_of_charge` :arrow_right: `_battery_storage_state_of_charge` (and remove `_li_ion`)
https://github.com/TUMCREATE-ESTL/cobmo/compare/feature/thermal_storage#diff-bbc55971cd554ab8d16f6643b1f6b422R133
- [ ] Split up to sensible / latent to enable having multiple storage at the same time: `_storage_charge_heat_thermal_power` :arrow_right: `_sensible_storage_charge_heat_thermal_power`  and so on
https://github.com/TUMCREATE-ESTL/cobmo/compare/feature/thermal_storage#diff-bbc55971cd554ab8d16f6643b1f6b422R176
- [ ] `_storage_charge_electric_power` :arrow_right: `_battery_storage_charge_electric_power`
https://github.com/TUMCREATE-ESTL/cobmo/compare/feature/thermal_storage#diff-bbc55971cd554ab8d16f6643b1f6b422R133
- [ ] Same as above for outputs
https://github.com/TUMCREATE-ESTL/cobmo/compare/feature/thermal_storage#diff-bbc55971cd554ab8d16f6643b1f6b422R247
- [ ] Same as above in `define_output_hvac_ahu_electric_power` and `define_output_hvac_tuu_electric_power`

## General Comments

- [ ] Avoid putting temporary CSV files in the repo, e.g. `cobmo/building___control_output_matrix.csv`

**Comment by tommasomiori** [comment]
- Since the heating and cooling efficiencies are the same for sensible and latent, like this I am defining the two techs with the same command. Can work?

https://github.com/TUMCREATE-ESTL/cobmo/commit/603d2de45962e34f9ba0c43799bbf1ace4ea7a1a#diff-bbc55971cd554ab8d16f6643b1f6b422R2381

- I see you add again the value in the same matrix cell.
`self.control_output_matrix.at[
                            index + '_tu_heat_electric_power',
                            index + '_storage_to_zone_heat_power'
                        ]`
Is this to be sure that if something was there before you take it into account?
https://github.com/TUMCREATE-ESTL/cobmo/commit/603d2de45962e34f9ba0c43799bbf1ace4ea7a1a#diff-bbc55971cd554ab8d16f6643b1f6b422R2482

**Comment by stroitzsch** [comment]
> Since the heating and cooling efficiencies are the same for sensible and latent, like this I am defining the two techs with the same command. Can work?
> 603d2de#diff-bbc55971cd554ab8d16f6643b1f6b422R2381

But when we have 1 sensible and 1 latent storage, we will need to have two separate heat transfer variables right? And hence, separate entry in output control matrix.

> I see you add again the value in the same matrix cell.
> ... = self.control_output_matrix.at[ index + '_tu_heat_electric_power', index + '_storage_to_zone_heat_power' ] + ... actual part in the matrix
> Is this to be sure that if something was there before you take it into account?
> 603d2de#diff-bbc55971cd554ab8d16f6643b1f6b422R2482

Yes, it's just to be safe. Probably not needed most of the time, but it's just safer.

**Comment by stroitzsch** [comment]
FYI: I moved the storage type definition from `building_scenarios` to `buildings`, but it should still work as expected.
https://github.com/TUMCREATE-ESTL/cobmo/blob/c5937e565eb873a8955f4ca3e49b96b1a744203d/cobmo/database_schema.sql#L166

**Comment by tommasomiori** [comment]
> > Since the heating and cooling efficiencies are the same for sensible and latent, like this I am defining the two techs with the same command. Can work?
> > [603d2de](https://github.com/TUMCREATE-ESTL/cobmo/commit/603d2de45962e34f9ba0c43799bbf1ace4ea7a1a)#diff-bbc55971cd554ab8d16f6643b1f6b422R2381
> 
> But when we have 1 sensible and 1 latent storage, we will need to have two separate heat transfer variables right? And hence, separate entry in output control matrix.
> 
> > I see you add again the value in the same matrix cell.
> > ... = self.control_output_matrix.at[ index + '_tu_heat_electric_power', index + '_storage_to_zone_heat_power' ] + ... actual part in the matrix
> > Is this to be sure that if something was there before you take it into account?
> > [603d2de](https://github.com/TUMCREATE-ESTL/cobmo/commit/603d2de45962e34f9ba0c43799bbf1ace4ea7a1a)#diff-bbc55971cd554ab8d16f6643b1f6b422R2482
> 
> Yes, it's just to be safe. Probably not needed most of the time, but it's just safer.

I have just pushed these changes

**Comment by tommasomiori** [comment]
In the last commit, I have updated the AHU and TU output definitions.
- I have realized that I had defined too many variables is the output vector.
- I was not accounting for charge in the electric power of AHU and TU

There is a **remark** here:
I am making the **assumption** that the charge and discharge of the TES (sensible and latent) are accounted for the AHU just in terms of thermal power that the unit has to provide in addition (charge) or less (discharge). In reality, I imagine that the thermal storage would help only the cooling coil in the AHU and not the whole unit. 
Referring to the image, the storage is helping only the E --> ADP transformation. 
![AHU cooling psychometric](https://user-images.githubusercontent.com/26601099/62025826-d043cf00-b20b-11e9-936c-bb972e27be89.PNG)

**Comment by tommasomiori** [comment]
@stroitzsch 
- I defined almost everything as a parameter in the _parameter_sets_. Is this all right or it's better to do it in the building_stare_types?

- I can't really recognize if this will have an effect. After all, it will iterate along 1 item (building) only, correct? (https://github.com/TUMCREATE-ESTL/cobmo/commit/dfacd794bc7234486d618d7246c68069439255f1#diff-bbc55971cd554ab8d16f6643b1f6b422R424)

**Comment by stroitzsch** [comment]
> I defined almost everything as a parameter in the parameter_sets. Is this all right or it's better to do it in the building_stare_types?
- [ ] Do it in `building_storage_types` instead. The parameters table is an additional layer for easily changing parameters centrally, but should shadow the type tables.
> I can't really recognize if this will have an effect. After all, it will iterate along 1 item (building) only, correct? (dfacd79#diff-bbc55971cd554ab8d16f6643b1f6b422R424)
- [ ] No, iterating over the building won't be necessary.

**Comment by tommasomiori** [comment]
Can it work now calling the parameters like:
`self.building_scenarios['_parameter_name_']`

With the parameters defined in `building_storage_types `

**Comment by tommasomiori** [comment]
> Can it work now calling the parameters like this?
> `self.building_scenarios['_parameter_name_']`
> 
> With the parameters defined in `building_storage_types `
> 
> The commit is pretty messy because I added an IF clause ([bed2798#diff-bbc55971cd554ab8d16f6643b1f6b422R425](https://github.com/TUMCREATE-ESTL/cobmo/commit/bed2798cf05685e0f34873584d1d1b723004cfce#diff-bbc55971cd554ab8d16f6643b1f6b422R425))

Just add parse parameter everywhere
`parse_parameter(self.building_scenarios['storage_UA_external'])`

**Comment by tommasomiori** [comment]
> > Can it work now calling the parameters like this?
> > `self.building_scenarios['_parameter_name_']`
> > With the parameters defined in `building_storage_types `
> > The commit is pretty messy because I added an IF clause ([bed2798#diff-bbc55971cd554ab8d16f6643b1f6b422R425](https://github.com/TUMCREATE-ESTL/cobmo/commit/bed2798cf05685e0f34873584d1d1b723004cfce#diff-bbc55971cd554ab8d16f6643b1f6b422R425))
> 
> Just add parse parameter everywhere
> `self.parse_parameter(self.building_scenarios['storage_UA_external'])`

Nice, thanks man!

**Comment by tommasomiori** [comment]
> > > Can it work now calling the parameters like this?
> > > `self.building_scenarios['_parameter_name_']`
> > > With the parameters defined in `building_storage_types `
> > > The commit is pretty messy because I added an IF clause ([bed2798#diff-bbc55971cd554ab8d16f6643b1f6b422R425](https://github.com/TUMCREATE-ESTL/cobmo/commit/bed2798cf05685e0f34873584d1d1b723004cfce#diff-bbc55971cd554ab8d16f6643b1f6b422R425))
> > 
> > 
> > Just add parse parameter everywhere
> > `self.parse_parameter(self.building_scenarios['storage_UA_external'])`
> 
> Nice, thanks man!

No problem bro ;)

**Comment by tommasomiori** [comment]
I have defined the storage constraints.
Running the main gives an error in line 24 of `main.py` (https://github.com/TUMCREATE-ESTL/cobmo/blob/feature/thermal_storage/cobmo/main.py#L24) on opening the _sqlite_ in `utils.py` (https://github.com/TUMCREATE-ESTL/cobmo/blob/feature/thermal_storage/cobmo/utils.py#L36).
ERROR:
`cursor.executescript(open(sql_path, 'r').read())
sqlite3.OperationalError: near ")": syntax error`

**Comment by stroitzsch** [comment]
Check the `database_schema.sql` file. - Sounds like there might be an error there.

**Comment by tommasomiori** [comment]
Have you ever seen this error related to interpolation? (https://github.com/TUMCREATE-ESTL/cobmo/blob/feature/thermal_storage/cobmo/building.py#L3101)
`TypeError: Cannot interpolate with all object-dtype columns in the DataFrame. Try setting at least one column to a numeric dtype.`

**Comment by stroitzsch** [comment]
> Have you ever seen this error related to interpolation? (https://github.com/TUMCREATE-ESTL/cobmo/blob/feature/thermal_storage/cobmo/building.py#L3101)
> TypeError: Cannot interpolate with all object-dtype columns in the DataFrame. Try setting at least one column to a numeric dtype.
Fixed.

**Comment by tommasomiori** [comment]
With the latest commit, the code runs with no errors. (6e215a1596afdf25870e8a706117e9adc4752179)
The controller does not work though. Outputs are `NaN`

**Comment by tommasomiori** [comment]
> * also, the initial value of  `state_timeseries_simulation ` is 26 instead of 0.0 as defined.
> * the `create_level_8_4a_sensible_thermal_storage_mass_factor` has a value of 26 for all the timesteps

- [x] Solved by changing the order of the initialization in both `main.py` and `controller.py`  (https://github.com/TUMCREATE-ESTL/cobmo/commit/a4bf9377ed9b7807b1233668cec5fa7ba4285316#diff-f1bdbaf674d18ed6aa85d16d5d9d2cadR121-R124)

**Comment by tommasomiori** [comment]
> > * the `create_level_8_4a_sensible_thermal_storage_mass_factor` has a value of 26 for all the timesteps
> 
> ❌ still persisting

Got rid of the mass_factor at all ==> No losses. (https://github.com/TUMCREATE-ESTL/cobmo/blob/feature/thermal_storage/cobmo/building.py#L425-L447)

**Comment by tommasomiori** [comment]
> * added lower boundary for charge and discharge heat flows, but the constraint has no effect. ([708b9af](https://github.com/TUMCREATE-ESTL/cobmo/commit/708b9afb81a9da4cac6c699512395951977a11de)).
>   _The problem is that this constraint is for outputs, and I need to constraint a **contro**_

@stroitzsch do you think it would make sense to add some `rule_contraint_bound `here(https://github.com/TUMCREATE-ESTL/cobmo/blob/feature/thermal_storage/cobmo/controller.py#L150) as it's done for `rule_output_bound`? In this case, I would need to define it in the` building.py`

**Comment by tommasomiori** [comment]
> @stroitzsch do you think it would make sense to add some `rule_contraint_bound `here (https://github.com/TUMCREATE-ESTL/cobmo/blob/feature/thermal_storage/cobmo/controller.py#L150) as it's done for `rule_output_bound`? In this case, I would need to define it in the` building.py`

tried, but there's error pointing to controller (https://github.com/TUMCREATE-ESTL/cobmo/commit/9214f4ef62f85e6651f527b71484eba93ad79d65#diff-f1bdbaf674d18ed6aa85d16d5d9d2cadR103)

**Comment by tommasomiori** [comment]
Running with the storage is infeasible.
Notes:
- in the control_timeseries_controller it is visible that the storage is charging and discharging both cooling and heating, which doesn't make sense.

**Comment by stroitzsch** [comment]
In commit 02f983388c1f1009706be74fae862d44b410bc I made some `FIXME` comments that you should go through. I believe those should fix the remaining issues.

Feel free to comment here if unclear or disagree!

**Comment by tommasomiori** [comment]
- [ ] Although the outputs are defined here, the definition of the output matrix values is missing. See for example define_output_zone_temperature()
- [ ] separate output variable for electric power due to charging. like `building_ahu_storage_charge_electric_power` and `building_tu_storage_charge_electric_power`

- [ ] Move storage size to storage_types.

**Comment by tommasomiori** [comment]
> * [x]  Although the outputs are defined here, the definition of the output matrix values is missing. See for example define_output_zone_temperature()

new definitions: `define_output_storage_charge` and `define_output_storage_discharge`

**Comment by tommasomiori** [comment]
> * [ ]  Move storage size to storage_types.



> * [ ]  Move storage size to storage_types.

Maybe I'm not getting what you mean here, but I have defined the max capacity (i.e. the size) in teh `constraint_profiles `already.

**Comment by tommasomiori** [comment]
When minimizing `electric_power `instead of `power `without storage, the result is quite strange. The objective is lower, but the temperature is squeezed on the lower boundary. (https://github.com/TUMCREATE-ESTL/cobmo/commit/8ca18436a048f2655fedf8e04bd8c778fa40ee03#diff-f1bdbaf674d18ed6aa85d16d5d9d2cadR42)

Looking at the `output_timeseries_controller `I see that the reason is that the optimization can use cool power "for free" outside the constrained hours and cool down as much as possible. But the `cool_power `and the `electric_power `for AHU and TU should be connected so that this cannot happen, no?

**Comment by tommasomiori** [comment]
- [ ] put storage size into storage_type

- [ ] remove generic_heating : remove from the table

- [ ] electricity prices x electric_power into the controller.py

**Comment by tommasomiori** [comment]
**WORKING VERSION**

The last commit (88670b25d21ec532e2a21b17383d26c9c87bb4be) looks to work when comparing the results for `sensible_thermal_storage_default `on and off for `create_level_8_4a`.
What is different than before in the optimization is:
- the `electric_power `is subject to the optimization, and not the `power` in general.
- the electricity costs are included so that the objective function aims to minimize the cost and not the power.

**What was adjusted that made it work:**
It looks like the problem was with the noise creation in the state matrix (https://github.com/TUMCREATE-ESTL/cobmo/commit/88670b25d21ec532e2a21b17383d26c9c87bb4be#diff-bbc55971cd554ab8d16f6643b1f6b422R3521-R3523). In fact, without putting some noise, the matrix was having determinant zero and was hence a single matrix, which is not invertible. _I don't know why this was happening with the storage case but not with the normal one_. Anyway, the noise that was imposed was too big, as it was of the same magnitude of the order of the actual numbers related to the storage in the `state_matrix `(and these are really small since they come from the division 1/(cp*deltaT) = 1/(4186*8) which makes a pretty small number already.


**I have an observation related to the integration of this version into CONCEPT:**
Although I haven't checked this in CONCEPT yet, I imagine that somewhere the electricity price is multiplied to the electric power.  But I see that still what is optimized by the controller is the `power `and not the `electric_power`. For sure when I run CONCEPT with the storage COBMO I have to go back to a version without the electricity prices inside the COBMO, but I am not sure how it then will behave. In fact, as I saw, when you have storage and you don't use electricity prices in the optimization (controller.py) it is like having "free cooling". There might then be a problem in implementation somewhere. What do you think about this @stroitzsch?

**Comment by stroitzsch** [comment]
> It looks like the problem was with the noise creation in the state matrix (88670b2#diff-bbc55971cd554ab8d16f6643b1f6b422R3521-R3523). In fact, without putting some noise, the matrix was having determinant zero and was hence a single matrix, which is not invertible. I don't know why this was happening with the storage case but not with the normal one. Anyway, the noise that was imposed was too big, as it was of the same magnitude of the order of the actual numbers related to the storage in the state_matrix (and these are really small since they come from the division 1/(cp x deltaT) = 1/(4186 x 8) which makes a pretty small number already.
This is quite interesting but I don't really understand why it happens, because it shouldn't need to invert the matrix for anything? - Anyway, as long as it works, just keep it like this.

> When you look at the output_timeseries_controller for storage it seems like no charge is done (see the last column).To ease the comparison, I am creating slack files with all the matrixes into two folders, delete_me and delete_me_storage.
Did you already add the time-depended electricity price?

> Although I haven't checked this in CONCEPT yet, I imagine that somewhere the electricity price is multiplied to the electric power. But I see that still what is optimized by the controller is the power and not the electric_power. For sure when I run CONCEPT with the storage COBMO I have to go back to a version without the electricity prices inside the COBMO, but I am not sure how it then will behave. In fact, as I saw, when you have storage and you don't use electricity prices in the optimization (controller.py) it is like having "free cooling". There might then be a problem in implementation somewhere. 
I thought we talked about this: There is an error in the generic heating/cooling system where there is no electric power entry generated. So in the end it should be the electric power. You can just remove the generic system if you wanna make it work for now!

I'm not sure what you mean when comparing CONCEPT and COBMO, but I don't see any integration problems for now.

**Comment by tommasomiori** [comment]
In the commit (607fb87eb6e94615ab9574351a09927a829d3522) I noticed that there is a problem with the storage level. The charging mechanism is not working.

The problem might be connected with the control matrix, which looks to contain wrong values for the storage `state_of_charge`

**Comment by tommasomiori** [comment]
**NEW WORKING VERSION FOR STORAGE (ebdb3430a5a41102b3f8e7e36659e341332d64f3)**

What was creating the problem before was the "singular matrix problem". In fact, the workaround of the _noise_ was spoiling the results. Opening the state matrix is was possible to see that the matrix was not invertible because a 0 was set for the cell @ (`...state_of_charge...`, `...state_of_charge...`). 
To overcome this, a small but real loss is added as negatively contributing to the mass of the bottom layer. This fills in the empty cell and does not have a big influence on the storage behavior. 

**Changes in the approach**
- the charge is done only by the AHU. This is done because we assume that the AHU is the only one able to provide actual cold water. The TU provides only cold air. So said, the TU is able to receive cold water, therefore it can benefit from the storage discharge.

**Limitations**
- the storage is "helping" only the air handling unit

**To be further implemented**

- [ ] differentiate the `storage_to_zone` between TU and AHU to evaluate more accurately the benefits of storage on the building.

**Comment by tommasomiori** [comment]
last commit working for storage: (33a06c2d9cb81a9f54c77417289294947b486efd)
notes:
- look at `main_storage`
- in the objective function, the operational costs are considered per 15 years. meaning that the Power*price is multiplied times 14 levels * 260 working days per year * 15 years. 15 because this is the max timeline of thermal storage. This is the setup to have some installed by the optimization.
- at the end of the controller, the objective is brought back to the amount of SGD/day (still for 14 levels). This is done for easier plotting in the main
- no fixed capex (ie not depending on the storage size) is in the objective function.
- to make the plotting in the main, the reference number of running the combo without storage is pasted manually (https://github.com/TUMCREATE-ESTL/cobmo/commit/33a06c2d9cb81a9f54c77417289294947b486efd#diff-29da5b72ec991b55746139cd1b58bb4eR171). this is for 14 levels.

**Comment by tommasomiori** [comment]
discussion about _**how to divide the electric discharge from batteries into cooling coil and heating coil in the AHU**_
(https://github.com/TUMCREATE-ESTL/cobmo/blob/feature/thermal_storage/cobmo/building.py#L3035-L3094)

NOTE; do not consider the  `* (1 + 0.0000001)` since I have changed that.

**Comment by tommasomiori** [comment]
**BATTERY STORAGE** (daa3b9e2f3036f5f3711b009ba4014f854fe8824)
Moved the initial state definition inside the building.

- [ ] ERROR TO BE SOLVED STILL is the infeasibility that appears when setting the min battery state of charge depending on the DoD. For now, it is set to zero. Though this is a too big assumption to make. line = (https://github.com/TUMCREATE-ESTL/cobmo/commit/daa3b9e2f3036f5f3711b009ba4014f854fe8824#diff-bbc55971cd554ab8d16f6643b1f6b422R424-R426)

**Comment by tommasomiori** [comment]
CLEANING THE CODE
- [ ] delete the discharge and charge thermal powers from the output set
- [ ]

**Comment by tommasomiori** [comment]
### BESS - results @ 2019-08-30 (dbc38f5b59955cfed5ecf4ec5a0a59e0fbee0e59)

Economic Feasibility map for BESS

[best_plot_2019-08-30_11-26-57.pdf](https://github.com/TUMCREATE-ESTL/cobmo/files/3558190/best_plot_2019-08-30_11-26-57.pdf)

**Comment by tommasomiori** [comment]
TODOs from 
- differentiate utils in pLOTS and normal
- all mains need to run
- additional csv fles: those that store results and you read them later, put in the "results" folder. the files that are just needed to run the battery cases put them in the" data_sotrage"
- go through the changes in the building model.

**Comment by stroitzsch** [comment]
Closed by e004f1112605f465d6e7b66397fb426f3f4719d2

**Comment by stroitzsch** [comment]
Closed by e004f1112605f465d6e7b66397fb426f3f4719d2
